File: src/algorithms/dynamic-programming/can-jump/sources/can-jump.py
# Can Jump tabulation — determine if you can reach the last index from index 0
def can_jump(nums: list[int]) -> bool:  # @step:initialize
    table_size = len(nums)  # @step:initialize
    dp_table = [0] * table_size  # @step:initialize,fill-table
    dp_table[0] = 1  # @step:fill-table
    # For each index, check if any prior reachable index can reach it
    for target_index in range(1, table_size):  # @step:compute-cell
        for source_index in range(target_index):  # @step:read-cache
            if dp_table[source_index] == 1 and source_index + nums[source_index] >= target_index:  # @step:read-cache,compute-cell
                dp_table[target_index] = 1  # @step:compute-cell
                break
    return dp_table[table_size - 1] == 1  # @step:complete

File: src/algorithms/dynamic-programming/subsequence/can-jump/sources/can-jump.ts
// Can Jump tabulation — determine if you can reach the last index from index 0
function canJump(nums: number[]): boolean {
  // @step:initialize
  const tableSize = nums.length; // @step:initialize
  const dpTable = new Array(tableSize).fill(0); // @step:initialize,fill-table
  dpTable[0] = 1; // @step:fill-table
  // For each index, check if any prior reachable index can reach it
  for (let targetIndex = 1; targetIndex < tableSize; targetIndex++) {
    // @step:compute-cell
    for (let sourceIndex = 0; sourceIndex < targetIndex; sourceIndex++) {
      // @step:read-cache
      if (dpTable[sourceIndex] === 1 && sourceIndex + nums[sourceIndex] >= targetIndex) {
        // @step:read-cache,compute-cell
        dpTable[targetIndex] = 1; // @step:compute-cell
        break;
      }
    }
  }
  return dpTable[tableSize - 1] === 1; // @step:complete
}
